rfc/java11.md

```diff
@@ -0,0 +1,164 @@
+# Java 11 Migration
+
+| | |
+|-|-|
+| **Status**     | _**Proposed**, Accepted, Implemented, Obsolete_ |
+| **RFC #**      | [62](https://github.com/spinnaker/community/pull/62) |
+| **Author(s)**  | Michael Plump ([`@plumpy`](https://github.com/plumpy)) |
+| **SIG / WG**   | Platform SIG |
+
+## Overview
+
+To keep current with the evolution of Java, Spinnaker needs to move from Java 8
+to Java 11. This RFC documents a process and timeline for doing so.
+
+### Goals and Non-Goals
+
+Goals:
+
+*   Provide a process for moving official Spinnaker builds to the Java 11 JRE
+*   Discuss ways we will encourage people using non-offical builds to also move
+    to the Java 11 JRE
+*   Provide a timeline for the final switch to a Java 11 JDK along with criteria
+    to decide if the timeline should be delayed
+
+Non-Goals:
+
+*   Providing a rationale for upgrading Java; all active Java projects need to
+    update their JDK version eventually as libraries we depend upon will also
+    depend on new versions of the JDK
+*   Debating which Java version we should upgrade to; we take it as a given that
+    only LTS releases are acceptable for most organizations, especially for
+    running critical infrastructure like Spinnaker
+*   Documenting and solving technical issues that present themselves in the
+    migration; those will be handled separately, but this document concerns
+    itself primarily with the larger processes and timeline
+*   Detailing the required release system changes; since the releases are still
+    entirely maintained by Google, these changes aren't particularly relevant to
+    the wider community
+
+### JRE vs. JDK
+
+There are two components of Java of interest to this RFC. The Java Runtime
+Environment (JRE) is the `java` binary and associated libraries used to _run_
+Spinnaker services. The Java Development Kit (JDK) is the `javac` binary and
+associated libraries used to _compile_ Spinnaker.
+
+Once we upgrade to the Java 11 JDK, Spinnaker will no longer be able to run in a
+Java 8 JRE. So the first (more complicated) step is to get everyone migrated to
+a Java 11 JRE. Only after that is done can we switch to the Java 11 JDK.
+
+The Java 11 JRE has some fairly large changes, most notably the deprecation of
+the long-standing CMS garbage collector. Large organizations with highly-tuned
+JVMs have shown hesitancy to upgrade until the new runtime has undergone
+extensive testing. (Both Google and Netflix are in this category.)
+
+For official Spinnaker containers, we control the JRE used to run Spinnaker. For
+those releases, this proposal takes a relatively slow ramp to the Java 11 JRE
+and provides an escape hatch back to Java 8 for users who experience problems.
+
+There are also some unknown number of customers who aren't using the Spinnaker-
+provided containers. We will need to do some outreach to make sure these users
+convert to a Java 11 JRE too.
```

Comment:
Do we have an idea of who these customers are? Also, do you have a plan for how this outreach will happen?

Reply:
I think they're pretty much by definition going to be customers who are very familiar with Spinnaker, since they're doing their own builds and deployments without the benefit of Halyard. So I don't imagine there are really any that we don't already have contact with...

In addition to the release notes, announcement emails, and slack updates, I figured I would get a list of names from our team and yours and reach out directly (via email) to the ones we can think of...

Reply:
We use some "custom" builds but for the most part these are built off of the current builds that halyard pulls from anyways. So the upgrade process would be similar for us anyways.

Reply:
But are you building your own containers? The important thing here is that those containers would need to be switched to a Java 11 JRE, if they aren't already. I guess if you use the bundled Dockerfiles, this would be handled for you. (I'll just switch the first line of this file to openjdk 11: https://github.com/spinnaker/orca/blob/master/Dockerfile.slim)

Reply:
Yes we are building our own containers (to build in our own internal stuff into the services), but these are just references of the existing images that are published, so not a big deal for us.